geps/gep-3440/index.md

```diff
@@ -0,0 +1,283 @@
+    By default, Nginx only retries on network error and timeout,
+    but you can specify other conditions (like HTTP status codes) to expand retry options.
+2. **Retry Limit by Time** (`grpc_next_upstream_timeout`):
+   You can set a total time limit for how long Nginx will attempt retries.
+   This limits the retry process to a specified time window, after which Nginx will stop attempting further retries.
+3. **Retry Limit by Number** (`grpc_next_upstream_tries`):
+    You can set a maximum number of retry attempts for a request.
+    Once this limit is reached, Nginx will stop attempting further retries.
+4. **Non-Idempotent Requests** (`non_idempotent`):
+    By default, Nginx does not retry non-idempotent requests (like POST or PUT) because they can cause side effects
+    if sent multiple times. However, you can enable retries for non-idempotent requests if needed.
+
+**Important Considerations**:
+- **Partial Responses**: Nginx can only retry if no part of the response has been sent to the client.
+If an error occurs mid-response, retries are not possible.
+- **Unsuccessful Attempts**: Errors like `timeout` and `invalid_header` are always considered unsuccessful and will
+trigger retries if specified, while errors like `403` and `404` are not retryable by default.
+
+### HAProxy
+1. **Retry Conditions**: HAProxy can retry requests based on various network conditions
+(e.g., connection failures, timeouts) and some HTTP error codes. While HAProxy does support gRPC via HTTP/2, it does not
+have built-in support for handling specific gRPC status codes (like `Cancelled`, `Deadline Exceeded`).
+It relies on HTTP-level conditions for retries, so its gRPC support is less granular than the GEP requires.
+2. **Retry Limits**: HAProxy allows you to set a maximum number of retries for a request using the `retries` directive.
+It also supports setting a timeout for the entire retry process using the `timeout connect` and `timeout server` directives.
+
+### Traefik
+1. **Retry Conditions**: Traefik allows for retries based on HTTP-level conditions (e.g., connection errors and
+certain HTTP status codes like 500, 502, 503, and 504), but it does not natively interpret specific gRPC status codes
+like `UNAVAILABLE` or `DEADLINE_EXCEEDED`. This means that, while Traefik can retry requests on common HTTP errors
+that might represent temporary issues, it lacks the ability to directly handle and retry based on
+gRPC-specific error codes, limiting its alignment with the GEP’s requirement for granular gRPC status codes handling.
+2. **Retry Limits**: Traefik provides configurable retry attempts and can set a maximum number of retries. However,
+Traefik does not offer per-try timeout controls specific to each retry attempt. Instead, it typically relies on a
+global request timeout, limiting the flexibility needed for more precise gRPC retry management (like Envoy’s `per_try_timeout`).
+
```

Comment:
Linkerd supports gRPC retry as well: you MUST configure a GRPCRoute for Linkerd to understand that gRPC semantics are desired, but after that you can configure retries either on Routes or Services. See https://linkerd.io/2.17/reference/retries/.

Reply:
Thank you will add that!